File: 백준/Gold/11066. 파일 합치기/파일 합치기.py
T = int(input())

for _ in range(T):
    K = int(input())
    lst = [*map(int, input().split())]
    DP = [[0 for _ in range(K)] for _ in range(K)]

    pfs = [0] * (K + 1)
    for i in range(K):
        pfs[i + 1] = pfs[i] + lst[i]

    for n in range(2, K + 1):
        for i in range(K - n + 1):
            mini = float("inf")
            for j in range(i, i + n - 1):
                mini = min(mini, DP[i][j] + DP[j + 1][i + n - 1])
            DP[i][i + n - 1] = mini + (pfs[i + n] - pfs[i])

    print(DP[0][K - 1])

File: 백준/Gold/11066. 파일 합치기/파일 합치기.cc
#include <iostream>
#include <vector>
#include <algorithm>
#include <climits>

using namespace std;

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    int T;
    cin >> T;
    while (T--) {
        int K;
        cin >> K;

        vector<int> lst(K);
        vector<long long> pfs(K + 1, 0);
        for (int i = 0; i < K; ++i) {
            cin >> lst[i];
            pfs[i + 1] = pfs[i] + lst[i];
        }

        vector<vector<long long>> DP(K, vector<long long>(K, 0));

        for (int n = 2; n <= K; ++n) {
            for (int i = 0; i <= K - n; ++i) {
                int j = i + n - 1;
                long long mini = LLONG_MAX;

                for (int k = i; k < j; ++k) {
                    long long cost = DP[i][k] + DP[k + 1][j];
                    if (cost < mini) mini = cost;
                }
                DP[i][j] = mini + (pfs[j + 1] - pfs[i]);
            }
        }

        cout << DP[0][K - 1] << "\n";
    }

    return 0;
}